src/multiplex/mod.rs

```diff
@@ -35,85 +35,68 @@ mod client;
 mod multiplex;
 mod server;
 
-pub use self::multiplex::{Multiplex, Dispatch};
+pub use self::multiplex::{Multiplex, MultiplexMessage, Dispatch};
 pub use self::client::connect;
 pub use self::server::Server;
 
-use {Message};
+use {Error};
 use tokio_core::io::FramedIo;
-use tokio_service::{Service};
-use futures::{Async, Future, IntoFuture, Poll};
-use futures::stream::{Stream, Sender};
-use take::Take;
+use futures::{Async, Poll};
 use std::{fmt, io};
 
 /// Identifies a request / response thread
 pub type RequestId = u64;
 
 /// A multiplexed protocol frame
 pub enum Frame<T, B, E> {
-    /// Either a request or a response
-    Message(RequestId, T),
-    /// Returned by `Transport::read` when a streaming body will follow.
-    /// Subsequent body frames with a matching `RequestId` will be proxied to
-    /// the provided `Sender`.
-    ///
-    /// Calling `Transport::write` with Frame::MessageWithBody is an error.
-    MessageWithBody(RequestId, T, Sender<B, E>),
-    /// Body frame. None indicates that the body is done streaming.
-    Body(RequestId, Option<B>),
+    /// Either a request or a response.
+    Message {
+        /// Message exchange identifier
+        id: RequestId,
+        /// The message value
+        message: T,
+        /// Set to true when body frames will follow with the same request ID.
+        body: bool,
+    },
+    /// Body frame.
+    Body {
+        /// Message exchange identifier
+        id: RequestId,
+        /// Body chunk. Setting to `None` indicates that the body is done
+        /// streaming and there will be no further body frames sent with the
+        /// given request ID.
+        chunk: Option<B>,
+    },
     /// Error
-    Error(RequestId, E),
+    Error {
+        /// Message exchange identifier
+        id: RequestId,
+        /// Error value
+        error: E,
+    },
     /// Final frame sent in each transport direction
     Done,
 }
 
-/// A specialization of `Service` supporting the requirements of server
-/// pipelined services
-///
-/// `Service` should be implemented instead of this trait.
-pub trait ServerService {
-    /// Requests handled by the service.
-    type Request;
-
-    /// Responses given by the service.
-    type Response;
-
-    /// Response body chunk
-    type Body;
-
-    /// Response body stream
-    type BodyStream: Stream<Item = Self::Body, Error = Self::Error>;
-
-    /// Errors produced by the service.
-    type Error;
-
-    /// The future response value.
-    type Future: Future<Item = Message<Self::Response, Self::BodyStream>, Error = Self::Error>;
-
-    /// Process the request and return the response asynchronously.
-    fn call(&self, req: Self::Request) -> Self::Future;
-}
-
 /// A specialization of `io::Transport` supporting the requirements of
 /// pipeline based protocols.
 ///
 /// `io::Transport` should be implemented instead of this trait.
-pub trait Transport {
+pub trait Transport: 'static {
```

Comment:
Ah ok so you think that the multiple/pipeline specific `Transport` traits will want to stick around? In that case it may be worth brainstorming names here as having three `Transport` traits can get confusing, but we can take that to an issue as well